### PR TITLE
EfficientNet support

### PR DESCRIPTION
**Pull request by AIStatistics**
Support the EfficientNet model in the paper "EfficientNet: Rethinking Model Scaling for Convolutional Neural Networks" by Mingxing Tan, Quoc V. Le, 2019

### CODE REVIEW

File: dlpy/applications/__init__.py [hunk 1]
```diff
@@ -26,4 +26,5 @@
 from .shufflenet import ShuffleNetV1
 from .inception import InceptionV3
 from .rcnn import Faster_RCNN
-from .unet import UNet
\ No newline at end of file
+from .unet import UNet
+from .efficientnet import *
```
Comment (ethem-kinginthenorth) [inline dlpy/applications/__init__.py]: you dont need to import all, please just add the actual api that users will use here

File: dlpy/applications/efficientnet.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+    n_channels : int, optional
+        Specifies the number of the channels (i.e., depth) of the input layer.
+        Default: 3
+    width : int, optional
+        Specifies the width of the input layer.
+        Default: 224
+    height : int, optional
+        Specifies the height of the input layer.
+        Default: 224
+    width_coefficient: float,
```
Comment (ethem-kinginthenorth) [inline dlpy/applications/efficientnet.py]: can you please add the default values here

File: dlpy/applications/efficientnet.py [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+        Specifies the scale coefficient for network width.
+    depth_coefficient: float,
+        Specifies the scale coefficient for network depth.
+    dropout_rate: double, optional
+        Specifies the dropout rate before final classifier layer.
```
Comment (ethem-kinginthenorth) [inline dlpy/applications/efficientnet.py]: can you explain this a bit more

File: dlpy/applications/efficientnet.py [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+        if min_value is None:
+            min_value = divisor
+        new_v = max(min_value, int(v + divisor / 2) // divisor * divisor)
+        # make sure that round down does not go down by more than 10%.
+        if new_v < 0.9 * v:
+            new_v += divisor
+        return new_v
+
+    def round_filters(filters, width_coefficient, depth_divisor):
+        # round number of filters based on width multiplier and depth_divisor
```
Comment (ethem-kinginthenorth) [inline dlpy/applications/efficientnet.py]: can you add some more comments explaining where you use this

File: dlpy/applications/efficientnet.py [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+        filters *= width_coefficient
+        new_filters = int(filters + depth_divisor / 2) // depth_divisor * depth_divisor
+        new_filters = max(depth_divisor, new_filters)
+        # Make sure that round down does not go down by more than 10%.
+        if new_filters < 0.9 * filters:
+            new_filters += depth_divisor
+        return int(new_filters)
+
+    def round_repeats(repeats, depth_coefficient):
+        # round number of repeats based on depth multiplier
```
Comment (ethem-kinginthenorth) [inline dlpy/applications/efficientnet.py]: can you please add some more comment explaining where you use this

File: dlpy/applications/efficientnet.py [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+    :class:`Model`
+
+    References
+    ----------
+    https://arxiv.org/pdf/1905.11946.pdf
+
+    '''
+
+    def _make_divisible(v, divisor, min_value=None):
+        # make number of channel divisible
```
Comment (kesmit13) [inline dlpy/applications/efficientnet.py]: Even though these are internal methods, they should have docstrings indicating what the parameters are and the types.

File: dlpy/applications/efficientnet.py (hunk 4 above)
Comment (kesmit13) [inline dlpy/applications/efficientnet.py]: same as above

File: dlpy/applications/efficientnet.py [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+        return int(math.ceil(depth_coefficient * repeats))
+
+    def _MBConvBlock(inputs, in_channels, out_channels, ksize, stride, expansion, se_ratio, stage_id, block_id,
+                     noskip=False, activation_fn='relu'):
+        """
+        Inverted Residual Block
+
+        Parameters
+        ----------
+        inputs:
```
Comment (kesmit13) [inline dlpy/applications/efficientnet.py]: Add type information to parameters

Comment (ethem-kinginthenorth) [inline dlpy/applications/efficientnet.py]: please add documents for all the parameters

File: dlpy/applications/efficientnet.py [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+            Input tensor
+        in_channels:
+            Specifies the number of input tensor's channel
+        expansion:
```
Comment (ethem-kinginthenorth) [inline dlpy/applications/efficientnet.py]: in the order you provide in the API please

File: dlpy/applications/efficientnet.py [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+            expansion factor always applied to the input size.
+        stride:
+            the strides of the convolution
+        alpha:
```
Comment (ethem-kinginthenorth) [inline dlpy/applications/efficientnet.py]: I dont see any alpha parameter in the function

File: dlpy/applications/efficientnet.py [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+            width multiplier.
+        filters:
```
Comment (ethem-kinginthenorth) [inline dlpy/applications/efficientnet.py]: I dont see any filters parameter in the function

File: dlpy/applications/efficientnet.py [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+
+        if expansion > 1:
+            # For MobileNet V2, expansion>1 when stage>0
+            n_groups = int(expansion * in_channels)  ## update n_groups
+            x = Conv2d(n_groups, 1, include_bias=False, act='identity',
+                       name=prefix + 'expand')(x)
+            x = BN(name=prefix + 'expand_BN', act='identity')(x)
+
+        # Depthwise convolution with kernel size : 3 or 5
+        x = GroupConv2d(n_groups, n_groups, ksize, stride=stride, act='identity',
```
Comment (ethem-kinginthenorth) [inline dlpy/applications/efficientnet.py]: if ksize takes only 3 or 5 then you need to specify this in the docstring - maybe some checks here

File: dlpy/applications/efficientnet.py [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+    x = OutputLayer(n=n_classes)(x)
+
+    model = Model(conn, inp, x, model_table)
+    model.compile()
+    return model
+
+
+def EfficientNetB0(conn, model_table='EfficientNetB0',
+                   n_classes=1000,
+                   **kwargs):
```
Comment (ethem-kinginthenorth) [inline dlpy/applications/efficientnet.py]: why do you need kwargs?

File: dlpy/tests/test_applications.py [hunk 13]
```diff
@@ -913,3 +913,34 @@ def test_densenet121_onnx_2(self):
...
+
+        # input shape of fficientnet models
+        self.assertTrue(effnetb0.input_layers[0].output_size==(224,224,3))
+        self.assertTrue(effnetb1.input_layers[0].output_size==(240,240,3))
+        self.assertTrue(effnetb2.input_layers[0].output_size==(260,260,3))
+        self.assertTrue(effnetb3.input_layers[0].output_size==(300,300,3))
+        self.assertTrue(effnetb4.input_layers[0].output_size==(380,380,3))
+        self.assertTrue(effnetb5.input_layers[0].output_size==(456,456,3))
+        self.assertTrue(effnetb6.input_layers[0].output_size==(528,528,3))
+        self.assertTrue(effnetb7.input_layers[0].output_size==(600,600,3))
```
Comment (ethem-kinginthenorth) [inline dlpy/tests/test_applications.py]: we might need an extra line here

File: dlpy/applications/efficientnet.py [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+
+    x = GlobalAveragePooling2D(name="Global_avg_pool", dropout=dropout_rate)(x)
+    x = OutputLayer(n=n_classes)(x)
+
+    model = Model(conn, inp, x, model_table)
+    model.compile()
+    return model
+
+
+def EfficientNetB0(conn, model_table='EfficientNetB0',
```
Comment (kesmit13) [inline dlpy/applications/efficientnet.py]: Needs docstrings

File: dlpy/applications/efficientnet.py (hunk 2 above)
Comment (kesmit13) [inline dlpy/applications/efficientnet.py]: don't put commas at the end if there isn't an 'optional'

File: dlpy/applications/__init__.py (hunk 1 above)
Comment (AIStatistics) [reply dlpy/applications/__init__.py]: done

File: dlpy/applications/efficientnet.py (hunk 2 above)
Comment (AIStatistics) [reply dlpy/applications/efficientnet.py]: fixed

File: dlpy/applications/efficientnet.py (hunk 7 above)
Comment (AIStatistics) [reply dlpy/applications/efficientnet.py]: fixed it.

File: dlpy/applications/efficientnet.py (hunk 8 above)
Comment (AIStatistics) [reply dlpy/applications/efficientnet.py]: fixed it.

File: dlpy/applications/efficientnet.py (hunk 9 above)
Comment (AIStatistics) [reply dlpy/applications/efficientnet.py]: fixed

File: dlpy/applications/efficientnet.py (hunk 10 above)
Comment (AIStatistics) [reply dlpy/applications/efficientnet.py]: fixed

File: dlpy/applications/efficientnet.py (hunk 11 above)
Comment (AIStatistics) [reply dlpy/applications/efficientnet.py]: The kernel size is specified from the block's parameters, which can be different from 3 and 5. So I updated the comment.

File: dlpy/applications/efficientnet.py (hunk 12 above)
Comment (AIStatistics) [reply dlpy/applications/efficientnet.py]: The kwargs can be passed to the function call EfficientNet() within EfficientNetB0. 
The following example shows that the parameter "random_crop" in function EfficentNetB0 is used and passed to function EfficientNet:
effnetb0 = EfficientNetB0(s, model_table='EfficientNetB0', n_classes=10, random_crop='resizethencrop', random_flip='h')

File: dlpy/tests/test_applications.py (hunk 13 above)
Comment (AIStatistics) [reply dlpy/tests/test_applications.py]: done.

File: dlpy/applications/efficientnet.py (hunk 3 above)
Comment (AIStatistics) [reply dlpy/applications/efficientnet.py]: The three dimension for scaling in efficient are: width, depth, and resolution. Here depth_coefficient is for scaling depth.  The dropout_rate is the drop out rate in the network.

File: dlpy/applications/efficientnet.py (hunk 5 above)
Comment (AIStatistics) [reply dlpy/applications/efficientnet.py]: Round_repeats is for the scaling depth in efficientnet. I added more comments in the function doc.

File: dlpy/applications/efficientnet.py (hunk 3 above)
Comment (AIStatistics) [reply dlpy/applications/efficientnet.py]: done

File: dlpy/applications/efficientnet.py (hunk 4 above)
Comment (AIStatistics) [reply dlpy/applications/efficientnet.py]: done

File: dlpy/applications/efficientnet.py (hunk 6 above)
Comment (AIStatistics) [reply dlpy/applications/efficientnet.py]: fixed

File: dlpy/applications/efficientnet.py (hunk 4 above)
Comment (AIStatistics) [reply dlpy/applications/efficientnet.py]: fixed

File: dlpy/applications/efficientnet.py (hunk 7 above)
Comment (AIStatistics) [reply dlpy/applications/efficientnet.py]: fixed

File: dlpy/applications/efficientnet.py (hunk 14 above)
Comment (AIStatistics) [reply dlpy/applications/efficientnet.py]: fixed.

File: dlpy/applications/efficientnet.py [hunk 15]
```diff
@@ -0,0 +1,464 @@
...
+        Default: 1000
+
+    References
+    ----------
+    https://arxiv.org/pdf/1905.11946.pdf
+    '''
+
+    return EfficientNet(conn, model_table, n_classes, n_channels=3,
+                        width=224, height=224, width_coefficient=1, depth_coefficient=1, dropout_rate=0.2,
+                        **kwargs)
```
Comment (ethem-kinginthenorth) [inline dlpy/applications/efficientnet.py]: I dont find kwargs as a good fit for this content, I let it only for test purposes. Please update your code to reflect actual parameters

Comment (AIStatistics) [reply dlpy/applications/efficientnet.py]: fixed.

**Comment by AIStatistics** [comment]
Thanks!

**Pull request merged**